### PR TITLE
Additional metrics 

### LINKED ISSUE #108: Additional Metrics
**Is this a BUG REPORT, PERFORMANCE REPORT or FEATURE REQUEST?**:

Feature Request

Need additional metrics exposed via discoverer:

- upstream-services{cluster, tenant/namespace} -- (Number of upstream services, labeled by cluster and tenant/namespace)
- replicated-services{cluster, tenant/namespace} -- (Number of services replicated to gimbal cluster labeled by cluster and tenant/namespace)
- invalid-services{cluster, tenant/namespace} -- (Number of services unable to be replicated labeled by cluster and tenant/namespace)
- upstream-endpoints{cluster, tenant/namespace,service} -- (Number of endpoints - meaning IP:Port - in the upstream labeled by cluster, namespace, and service name)
- replicated-endpoints{cluster, tenant/namespace,service} -- (Number of endpoints - meaning IP:Port replicated to Gimbal cluster labeled by cluster, namespace and service name)
- invalid-endpoints{cluster, tenant/namespace,service} -- (Number of endpoints - meaning IP:Port - unable to be replicated to gimbal labeled by namespace and service name)

### PR DESCRIPTION
**Pull request by stevesloka**
Add additional metrics to the Discoverers:
- upstream-services
- replicated-services
-  invalid-services
- upstream-endpoints
- replicated-endpoints
- invalid-endpoints

// Fixes #108 
Signed-off-by: Steve Sloka <steves@heptio.com>

### CODE REVIEW

**Comment by rosskukulinski** [comment]
Additional detail: I believe `gimbal_discoverer_upstream_endpoints_total` is showing currently displaying the number of Endpoint objects, not the sum of endpoints in the Endpoint objects.

`gimbal_discoverer_replicated_endpoints_total` appears to be correct`

File: discovery/pkg/openstack/reconciler.go [hunk 1]
```diff
@@ -154,7 +154,22 @@ func (r *Reconciler) reconcile() {
...
+		// Log upstream services prometheus
+		r.Metrics.DiscovererUpstreamServicesMetric(r.BackendName, projectName, totalUpstreamServices)
+		r.Metrics.DiscovererInvalidServicesMetric(r.BackendName, projectName, totalUpstreamServices-len(loadbalancers))
+		r.Metrics.DiscovererUpstreamEndpointsMetric(r.BackendName, projectName, r.sumEndpoints(desiredEndpoints))
+
+		endpoints, err := r.GimbalKubeClient.CoreV1().Endpoints(projectName).List(metav1.ListOptions{})
+		if err != nil {
+			r.Logger.Error(err)
+		}
+		r.Metrics.DiscovererReplicatedEndpointsMetric(r.BackendName, projectName, r.sumEndpoints(endpoints.Items))
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: Is this competing with the metric being set in `sync.Action`? It's unclear to me why we need them in both places. Also seems like we are missing the equivalent for `DiscovererReplicatedServicesMetric`?

File: discovery/pkg/openstack/reconciler.go [hunk 2]
```diff
@@ -154,7 +154,22 @@ func (r *Reconciler) reconcile() {
...
 
 		desiredEndpoints := kubeEndpoints(r.BackendName, projectName, loadbalancers, pools)
+		for _, ep := range desiredEndpoints {
+			desiredEndpoints = append(desiredEndpoints, ep)
+		}
 		r.reconcileEndpoints(desiredEndpoints, currentEndpoints.Items)
+
+		// Log upstream services prometheus
+		r.Metrics.DiscovererUpstreamServicesMetric(r.BackendName, projectName, totalUpstreamServices)
+		r.Metrics.DiscovererInvalidServicesMetric(r.BackendName, projectName, totalUpstreamServices-len(loadbalancers))
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: nit: took me some time to figure out how this subtraction results in the number of invalid services. Can we create a variable `invalidServicesCount` or something and move it up closer to `totalUpstreamServices` and `loadbalancers`?

File: discovery/pkg/openstack/reconciler.go [hunk 3]
```diff
@@ -154,7 +154,22 @@ func (r *Reconciler) reconcile() {
...
+		r.Metrics.DiscovererUpstreamEndpointsMetric(r.BackendName, projectName, r.sumEndpoints(desiredEndpoints))
+
+		endpoints, err := r.GimbalKubeClient.CoreV1().Endpoints(projectName).List(metav1.ListOptions{})
+		if err != nil {
+			r.Logger.Error(err)
+		}
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: Do we need an else here to avoid using `endpoints` when `err == nil`?

File: discovery/pkg/sync/service.go [hunk 4]
```diff
@@ -69,7 +71,25 @@ func (action serviceAction) Sync(kubeClient kubernetes.Interface, metrics localm
...
+	// Log Total Services Metric
+	totalServices, err := getTotalServicesCount(kubeClient, action.ObjectMeta().GetNamespace(), backendName)
+	if err != nil {
+		logger.Error("Error getting total services count: ", err)
+	} else {
+		metrics.DiscovererReplicatedServicesMetric(backendName, action.service.GetNamespace(), totalServices)
+	}
+
+	// // Log Total Services Metric
+	// totalUpstreamServices, err := getTotalServicesCount(kubeClient, action.ObjectMeta().GetNamespace(), backendName)
```
Comment (alexbrand) [inline discovery/pkg/sync/service.go]: remove?

**Comment by stevesloka** [comment]
@alexbrand I 100% believe we could re-do how metrics are implemented, this was my comment last time we were on the phone. I have them tied to events since that's the trigger for when something changes. 

Looking into the other comments/issues now.

**Comment by rosskukulinski** [comment]
I think modified value @stevesloka

**Comment by rosskukulinski** [comment]
@stevesloka  I lied, the `endpoint` metrics should have a label which is the original service/LB name.  The reason being it can be helpful to count the # endpoints available for a given Service across multiple backends.

File: discovery/pkg/k8s/controller.go [hunk 5]
```diff
@@ -139,6 +153,28 @@ func skipProcessing(name, namespace string) bool {
 	return false
 }
 
+func (c *Controller) writeServiceMetrics(svc *v1.Service) error {
+	upstreamServices, err := c.serviceLister.Services(svc.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		return err
```
Comment (alexbrand) [inline discovery/pkg/k8s/controller.go]: Should we log instead of return here? We are not handling it at the call site

File: discovery/pkg/metrics/metrics.go [hunk 6]
```diff
@@ -14,87 +14,150 @@
...
 					Name: DiscovererErrorTotal,
 					Help: "Number of errors that have occurred in the Discoverer",
 				},
-				[]string{"backendname", "errortype"},
+				[]string{"backendname", "errortype", "backendtype"},
+			),
+			DiscovererUpstreamServicesGauge: prometheus.NewGaugeVec(
+				prometheus.GaugeOpts{
+					Name: DiscovererUpstreamServicesGauge,
+					Help: "Total number of services in the upstream backend cluster",
```
Comment (alexbrand) [inline discovery/pkg/metrics/metrics.go]: "upstream backend" sounds redundant to me. Thoughts?

File: discovery/pkg/metrics/metrics.go [hunk 7]
```diff
@@ -109,65 +172,143 @@ func (d *DiscovererMetrics) RegisterPrometheus() {
...
+// DiscovererReplicatedEndpointsMetric records the total replicated endpoints
+func (d *DiscovererMetrics) DiscovererReplicatedEndpointsMetric(namespace, serviceName string, totalEp int) {
+	m, ok := d.metrics[DiscovererReplicatedEndpointsGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, namespace, serviceName, d.backendType).Set(float64(totalEp))
+	}
+}
+
+// DiscovererInvalidEndpointsMetric records the total replicated endpoints
+func (d *DiscovererMetrics) DiscovererInvalidEndpointsMetric(namespace, serviceName string, totalEp int) {
```
Comment (alexbrand) [inline discovery/pkg/metrics/metrics.go]: Should we get rid of this one?

File: discovery/pkg/metrics/metrics.go [hunk 8]
```diff
@@ -14,87 +14,150 @@
...
+				prometheus.GaugeOpts{
+					Name: DiscovererReplicatedEndpointsGauge,
+					Help: "Total number of endpoints replicated the backend cluster",
+				},
+				[]string{"backendname", "namespace", "servicename", "backendtype"},
+			),
+			DiscovererInvalidEndpointsGauge: prometheus.NewGaugeVec(
+				prometheus.GaugeOpts{
+					Name: DiscovererInvalidEndpointsGauge,
+					Help: "Total number of endpoints invalid endpoints that could not be replicaed from the backend cluster",
```
Comment (alexbrand) [inline discovery/pkg/metrics/metrics.go]: typo: "endpoints invalid endpoints"
but not sure if we need this metric anyways.

File: discovery/pkg/openstack/reconciler.go [hunk 9]
```diff
@@ -210,3 +219,19 @@ func (r *Reconciler) reconcileEndpoints(desired, current []v1.Endpoints) {
 		r.syncqueue.Enqueue(sync.DeleteEndpointsAction(&ep))
 	}
 }
+
+func (r *Reconciler) sumEndpoints(eps v1.Endpoints) int {
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: nit: can be func instead of method

File: discovery/pkg/openstack/reconciler.go [hunk 10]
```diff
@@ -210,3 +219,19 @@ func (r *Reconciler) reconcileEndpoints(desired, current []v1.Endpoints) {
...
+	total := 0
+	for _, ep := range eps.Subsets {
+		total += len(ep.Addresses)
+	}
+	return total
+}
+
+func (r *Reconciler) getListFromMap(mp map[string]v1.Endpoints) []v1.Endpoints {
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: nit: can be func instead of method

File: discovery/pkg/sync/endpoints.go [hunk 11]
```diff
@@ -54,63 +55,71 @@ func (action endpointsAction) ObjectMeta() *metav1.ObjectMeta {
...
 	}
 	if err != nil {
 		return fmt.Errorf("error handling %s: %v", action, err)
 	}
 
-	metrics.EndpointsEventTimestampMetric(action.endpoints.GetNamespace(), backendName, action.endpoints.GetName(), time.Now().Unix())
+	metrics.EndpointsEventTimestampMetric(action.endpoints.GetNamespace(), action.endpoints.GetName(), time.Now().Unix())
+
+	// TODO: Move to lister()
+	totalEps, err := metrics.GetTotalEndpointsCount(kubeClient, action.endpoints.GetNamespace(), action.endpoints.GetName())
```
Comment (alexbrand) [inline discovery/pkg/sync/endpoints.go]: I am not following why we need to GET the endpoints object again, if we already have it in `action.endpoints`?

File: discovery/pkg/metrics/metrics.go [hunk 12]
```diff
@@ -109,65 +172,143 @@ func (d *DiscovererMetrics) RegisterPrometheus() {
...
+// DiscovererInfoMetric records version information
+func (d *DiscovererMetrics) DiscovererInfoMetric(version string) {
+	m, ok := d.metrics[DiscovererInfoGauge].(*prometheus.GaugeVec)
+	if ok {
+		m.WithLabelValues(d.backendName, version, d.backendType).Set(1)
+	}
+}
+
+// GetTotalServicesCount returns the number of services in a namespace for the particular backend
+func (d *DiscovererMetrics) GetTotalServicesCount(kubeclient kubernetes.Interface, namespace string) (int, error) {
```
Comment (alexbrand) [inline discovery/pkg/metrics/metrics.go]: These funcs seem somewhat out of place in that they are unrelated to the metrics package. Wondering if it makes sense to move these calls to the k8s API into the sync pkg, which is where this is called and it already has a dependency on the k8s packages.

File: discovery/pkg/sync/endpoints.go (hunk 11 above)
Comment (stevesloka) [reply discovery/pkg/sync/endpoints.go]: Good point, before I had the metric sum all endpoints for a namespace which was different.

File: discovery/pkg/metrics/metrics.go (hunk 12 above)
Comment (stevesloka) [reply discovery/pkg/metrics/metrics.go]: I needed a good home for this but couldn't find one. Also, I need access to `backendName` to filter on labels. If it's moved from metrics will just need to make those params public.

Comment (alexbrand) [reply discovery/pkg/metrics/metrics.go]: Aaaaah I see. I think another option might be to add a `backendName` field to `endpointsAction` and `serviceAction`, and do `AddEndpointsAction(backendName, endpoints)`. The downside is that we have to plumb `backendName` again through a couple of layers. Not sure if it's worth it at this time.

Comment (stevesloka) [reply discovery/pkg/metrics/metrics.go]: I moved stuff around, just picked the backend name out of metrics, PTAL

File: discovery/pkg/metrics/metrics.go [hunk 13]
```diff
@@ -22,79 +22,139 @@ import (
...
+				prometheus.GaugeOpts{
+					Name: DiscovererReplicatedEndpointsGauge,
+					Help: "Total number of endpoints replicated the backend",
+				},
+				[]string{"backendname", "namespace", "servicename", "backendtype"},
+			),
+			DiscovererInvalidEndpointsGauge: prometheus.NewGaugeVec(
+				prometheus.GaugeOpts{
+					Name: DiscovererInvalidEndpointsGauge,
+					Help: "Total number of endpoints invalid endpoints that could not be replicaed from the backend",
```
Comment (alexbrand) [inline discovery/pkg/metrics/metrics.go]: typo: endpoints invalid endpoints

Comment (stevesloka) [reply discovery/pkg/metrics/metrics.go]: 👍 ...also fixed `replicated`

**Comment by stevesloka** [comment]
Yes right now new api call for each service CRUD, let me see if I can pipe through the lister, save us the call.

**Comment by stevesloka** [comment]
ok I've got the lister piped in, so no additional api requests are required for metrics.

**Comment by alexbrand** [comment]
Nice! Should I test this out on my end or can we merge?

**Comment by rosskukulinski** [comment]
something still isn't quite right, but now it's with the service metrics.

To test, I stopped both discoverers, deleted all services in the `kuard` namespace in the Gimbal cluster, then launched my newly-built discoverers from your last commit.

The following has been discovered:

```
NAME           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE       LABELS
blue-httpbin   ClusterIP   None         <none>        80/TCP     34s       app=httpbin,gimbal.heptio.com/backend=blue,gimbal.heptio.com/service=httpbin
blue-kuard     ClusterIP   None         <none>        8080/TCP   34s       gimbal.heptio.com/backend=blue,gimbal.heptio.com/service=kuard
green-kuard    ClusterIP   None         <none>        8080/TCP   34s       gimbal.heptio.com/backend=green,gimbal.heptio.com/service=kuard
```

yet here's the metrics gathered:

```

gimbal_discoverer_replicated_services_total{app="kubernetes-discoverer",backendname="blue",backendtype="kubernetes",cluster="k8s",instance="172.20.36.129:8080",job="kubernetes-pods",kubernetes_namespace="gimbal-discovery",kubernetes_pod_name="blue-kubernetes-discoverer-7b9999b95-rvfsc",namespace="kuard",pod_template_hash="365555651"} | 1
gimbal_discoverer_replicated_services_total{app="kubernetes-discoverer",backendname="blue",backendtype="kubernetes",cluster="k8s",instance="172.20.36.129:8080",job="kubernetes-pods",kubernetes_namespace="gimbal-discovery",kubernetes_pod_name="blue-kubernetes-discoverer-7b9999b95-rvfsc",namespace="nginx",pod_template_hash="365555651"} | 0
gimbal_discoverer_replicated_services_total{app="kubernetes-discoverer",backendname="green",backendtype="kubernetes",cluster="k8s",instance="172.20.41.62:8080",job="kubernetes-pods",kubernetes_namespace="gimbal-discovery",kubernetes_pod_name="green-kubernetes-discoverer-8496f87599-gz64x",namespace="kuard",pod_template_hash="4052943155"} | 0

```

**Comment by rosskukulinski** [comment]
And `gimbal_discoverer_*_endpoints_total` still has the combined `servicename` which has the backendname in it vs the original upstream service name.

**Comment by stevesloka** [comment]
I'm having trouble reproducing the service error where the counts are off. Could you port-forward to the discoverer on `8080` and double check the `/metrics` endpoint?

**Comment by rosskukulinski** [comment]
```
gimbal_discoverer_replicated_services_total{backendname="green",backendtype="kubernetes",namespace="kuard"} 0
gimbal_discoverer_upstream_services_total{backendname="green",backendtype="kubernetes",namespace="kuard"} 1
```

logs:

```
e="2018-06-08T22:12:05Z" level=info msg="Started workers"
time="2018-06-08T22:12:05Z" level=info msg="Successfully handled: add endpoints 'kuard/green-kuard'"
time="2018-06-08T22:12:05Z" level=info msg="Successfully handled: add service 'kuard/green-kuard'"
time="2018-06-08T22:42:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-08T22:42:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-08T23:12:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-08T23:12:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-08T23:42:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-08T23:42:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-09T00:12:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-09T00:12:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-09T00:42:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-09T00:42:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-09T01:12:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-09T01:12:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
time="2018-06-09T01:42:05Z" level=info msg="Successfully handled: update endpoints 'kuard/green-kuard'"
time="2018-06-09T01:42:05Z" level=info msg="Successfully handled: update service 'kuard/green-kuard'"
```

**Comment by rosskukulinski** [comment]
@stevesloka it's possible my local container build is screwed up.  If you push a build from your environment I can test that.

**Comment by stevesloka** [comment]
Try this image: `stevesloka/gimbal-discoverer:master`

**Comment by stevesloka** [comment]
> And gimbal_discoverer_*_endpoints_total still has the combined servicename which has the backendname in it vs the original upstream service name

I thought this was only for upstream services, not all. I can work to update.

**Comment by stevesloka** [comment]
@alexbrand @rosskukulinski I think this is ready to review again. The only things that I think could be improved are writing more tests around the Prom metrics. 

Additionally, there may be cases where we need to write out "0" for some error conditions. For example, should we always write a zero for gimbal_discoverer_invalid_services_total?

I have a prebuilt image here: `stevesloka/gimbal-discoverer:master`if that helps.

**Comment by rosskukulinski** [comment]
@stevesloka This is now working for me!  I don't think we need to write out a zero for metrics that don't have a value.

File: discovery/pkg/k8s/controller.go [hunk 14]
```diff
@@ -139,6 +155,18 @@ func skipProcessing(name, namespace string) bool {
 	return false
 }
 
+func (c *Controller) writeServiceMetrics(svc *v1.Service) {
+	upstreamServices, err := c.serviceLister.Services(svc.GetNamespace()).List(labels.Everything())
+	if err != nil {
+		c.Logger.Error("Could not get service metrics: ", err)
```
Comment (alexbrand) [inline discovery/pkg/k8s/controller.go]: I think we need a return after this

File: discovery/pkg/k8s/controller_test.go [hunk 15]
```diff
@@ -269,21 +234,114 @@ func TestUpdateEndpointsQueue(t *testing.T) {
...
+			c.writeEndpointsMetrics(tc.endpoint)
+
+			gatherers := prometheus.Gatherers{
+				metrics.Registry,
+				prometheus.DefaultGatherer,
+			}
+
+			gathering, err := gatherers.Gather()
+			if err != nil {
+				fmt.Println(err)
```
Comment (alexbrand) [inline discovery/pkg/k8s/controller_test.go]: t.Fatal here?

File: discovery/pkg/k8s/controller_test.go [hunk 16]
```diff
@@ -269,21 +234,114 @@ func TestUpdateEndpointsQueue(t *testing.T) {
...
+			c.writeServiceMetrics(tc.service)
+
+			gatherers := prometheus.Gatherers{
+				metrics.Registry,
+				prometheus.DefaultGatherer,
+			}
+
+			gathering, err := gatherers.Gather()
+			if err != nil {
+				fmt.Println(err)
```
Comment (alexbrand) [inline discovery/pkg/k8s/controller_test.go]: t.Fatal here?

File: discovery/pkg/openstack/reconciler.go [hunk 17]
```diff
@@ -117,48 +123,64 @@ func (r *Reconciler) reconcile() {
...
-			r.Metrics.GenericMetricError(r.BackendName, "ListEndpointsInNamespace")
+			r.Metrics.GenericMetricError("ListEndpointsInNamespace")
 			log.Errorf("error listing endpoints in namespace:%q: %v", projectName, err)
 			continue
 		}
 
+		// Convert the k8s list to type []Endpoints so make comparison easier
+		currentEndpoints := []Endpoints{}
+		for _, v := range currentk8sEndpoints.Items {
+			currentEndpoints = append(currentEndpoints, Endpoints{endpoints: v, upstreamName: ""})
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: Do we need to set the `upstreamName` here?

File: discovery/pkg/openstack/reconciler.go [hunk 18]
```diff
@@ -181,32 +203,34 @@ func (r *Reconciler) skipInvalidLoadBalancers(projectName string, lbs []loadbala
...
-		r.syncqueue.Enqueue(sync.UpdateEndpointsAction(&ep))
+	for _, ep := range del {
+		r.syncqueue.Enqueue(sync.DeleteEndpointsAction(&ep.endpoints, ep.upstreamName))
 	}
-	for _, e := range del {
-		ep := e
-		r.syncqueue.Enqueue(sync.DeleteEndpointsAction(&ep))
+}
+
+func getListFromMap(mp map[string]v1.Endpoints) []v1.Endpoints {
```
Comment (alexbrand) [inline discovery/pkg/openstack/reconciler.go]: Doesn't seem to be used anywhere

File: discovery/pkg/sync/endpoints_test.go [hunk 19]
```diff
@@ -138,7 +141,166 @@ func TestUpdateEndpoints(t *testing.T) {
...
+			a.LogMetrics(client, metrics, logrus.New())
+
+			gatherers := prometheus.Gatherers{
+				metrics.Registry,
+				prometheus.DefaultGatherer,
+			}
+
+			gathering, err := gatherers.Gather()
+			if err != nil {
+				fmt.Println(err)
```
Comment (alexbrand) [inline discovery/pkg/sync/endpoints_test.go]: t.Fatal?

File: discovery/pkg/sync/endpoints_test.go [hunk 20]
```diff
@@ -138,7 +141,166 @@ func TestUpdateEndpoints(t *testing.T) {
...
+				"backendname": backendName,
+				"backendtype": backendType,
+				"namespace":   "foo",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			nowFunc = func() time.Time {
```
Comment (alexbrand) [inline discovery/pkg/sync/endpoints_test.go]: How/where is this used?

File: discovery/pkg/sync/endpoints_test.go [hunk 21]
```diff
@@ -138,7 +141,166 @@ func TestUpdateEndpoints(t *testing.T) {
...
+	backendType := "backtype"
+	tests := []struct {
+		name              string
+		actionKind        string
+		endpoints         v1.Endpoints
+		existingendpoints v1.Endpoints
+		expectErr         bool
+		expectedCount     float64
+		expectedTimestamp float64
+		expectedLabels    map[string]string
```
Comment (alexbrand) [inline discovery/pkg/sync/endpoints_test.go]: Doesn't look like there are assertions on the labels

File: discovery/pkg/sync/queue.go [hunk 22]
```diff
@@ -110,17 +108,21 @@ func (sq *Queue) processNextWorkItem() bool {
...
-		sq.Metrics.QueueSizeGaugeMetric(sq.BackendName, sq.ClusterType, sq.Workqueue.Len())
+		sq.Metrics.QueueSizeGaugeMetric(sq.Workqueue.Len())
 		sq.Logger.Errorf("got an unknown item of type %T in the queue", obj)
 		return true
 	}
 
-	err := action.Sync(sq.KubeClient, sq.Metrics, sq.BackendName)
+	err := action.Sync(sq.KubeClient, sq.Logger)
+	if err != nil {
+		sq.Metrics.ServiceMetricError(action.ObjectMeta().GetNamespace(), action.ObjectMeta().GetName(), action.GetActionType())
```
Comment (alexbrand) [inline discovery/pkg/sync/queue.go]: What if we get an error while handling an Endpoints resource?

File: discovery/pkg/sync/endpoints.go [hunk 23]
```diff
@@ -28,90 +28,83 @@ import (
...
-	metrics.EndpointsEventTimestampMetric(action.endpoints.GetNamespace(), backendName, action.endpoints.GetName(), time.Now().Unix())
 	return nil
 }
 
 func (action endpointsAction) String() string {
 	return fmt.Sprintf(`%s endpoints '%s/%s'`, action.kind, action.endpoints.Namespace, action.endpoints.Name)
 }
 
-func addEndpoints(kubeClient kubernetes.Interface, endpoints *v1.Endpoints, lm localmetrics.DiscovererMetrics, backendName string) error {
+func (action endpointsAction) LogMetrics(gimbalKubeClient kubernetes.Interface, metrics localmetrics.DiscovererMetrics,
```
Comment (alexbrand) [inline discovery/pkg/sync/endpoints.go]: nit: Can we call this `SetMetrics` or `UpdateMetrics`? Log makes me think of log messages

File: discovery/pkg/sync/time.go [hunk 24]
```diff
@@ -0,0 +1,22 @@
+package sync
+
+import "time"
+
+type nowFuncT func() time.Time
+
+var nowFunc nowFuncT
+
+func init() {
```
Comment (alexbrand) [inline discovery/pkg/sync/time.go]: Can we add some docs on why we need this and what it is doing?

File: discovery/cmd/kubernetes-discoverer/main.go [hunk 25]
```diff
@@ -46,7 +46,7 @@ var (
 
 func init() {
 	flag.BoolVar(&printVersion, "version", false, "Show version and quit")
-	flag.IntVar(&numProcessThreads, "num-threads", 2, "Specify number of threads to use when processing queue items.")
+	flag.IntVar(&numProcessThreads, "num-threads", 1, "Specify number of threads to use when processing queue items.")
```
[the remark this answers is not in the document]
Comment (stevesloka) [reply discovery/cmd/kubernetes-discoverer/main.go]: nah, just local debugging, I'll restore

File: discovery/pkg/sync/endpoints_test.go (hunk 20 above)
Comment (stevesloka) [reply discovery/pkg/sync/endpoints_test.go]: The Timestamp metric used `time.Now()` to get the current time. In order to unit test that metric, we need a way to set it in the test. If it's not set in the test, then it defaults to now(). Both the service & endpoint in package sync use this.

File: discovery/pkg/sync/endpoints_test.go (hunk 21 above)
Comment (stevesloka) [reply discovery/pkg/sync/endpoints_test.go]: No, I didn't write tests for those. They are simple pass-through values. But initially, I had though to implement those.

File: discovery/pkg/sync/endpoints.go (hunk 23 above)
Comment (stevesloka) [reply discovery/pkg/sync/endpoints.go]: How about `WriteMetrics`? Set & Update sound like they are values that are getting set.

Comment (stevesloka) [reply discovery/pkg/sync/endpoints.go]: eh we can do `SetMetrics` I think

File: discovery/pkg/sync/time.go (hunk 24 above)
Comment (stevesloka) [reply discovery/pkg/sync/time.go]: Just inline comments? We don't have any dev docs at the moment.

File: discovery/pkg/sync/queue.go (hunk 22 above)
Comment (stevesloka) [reply discovery/pkg/sync/queue.go]: yup, meant to refactor that, good catch.

File: discovery/pkg/sync/time.go (hunk 24 above)
Comment (alexbrand) [reply discovery/pkg/sync/time.go]: Yep, sorry! That is what I meant :)

File: discovery/pkg/sync/endpoints_test.go (hunk 21 above)
Comment (alexbrand) [reply discovery/pkg/sync/endpoints_test.go]: Got it! Should we remove them from this struct?

File: discovery/pkg/sync/time.go (hunk 24 above)
Comment (stevesloka) [reply discovery/pkg/sync/time.go]: I cleaned it up, removed unused methods, added the missing header, and added some comments.

File: discovery/pkg/openstack/reconciler.go (hunk 17 above)
Comment (stevesloka) [reply discovery/pkg/openstack/reconciler.go]: Those are the current endpoints which come from the k8s query. The upstream name is not used, but having two lists of the same type makes the comparison logic work as-is.

File: discovery/pkg/sync/time.go [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+package sync
+
+import "time"
+
+// This is used to allow for unit testing values that rely on `time.Now()`.
+// In order to write tests and expect certain values, it's nessesary to provide
+// a way to override the default values.
+type nowFuncT func() time.Time
+
+var nowFunc nowFuncT
```
Comment (alexbrand) [inline discovery/pkg/sync/time.go]: I think this will result in a nil-pointer deref outside of tests

File: discovery/pkg/openstack/translate.go [hunk 27]
```diff
@@ -114,6 +114,15 @@ func serviceName(lb loadbalancers.LoadBalancer) string {
 	return strings.ToLower(lbName)
 }
 
+// get the lb Name or ID if name is empty
+func serviceNameOriginal(lb loadbalancers.LoadBalancer) string {
```
Comment (alexbrand) [inline discovery/pkg/openstack/translate.go]: Will this be a problem if there are two load balancers with the same name? Seems like it'll affect the metrics, but not 100% sure.

Comment (stevesloka) [reply discovery/pkg/openstack/translate.go]: Yes, it will. @rosskukulinski do you want to allow duplicates here? It only should affect openstack at the moment.

File: discovery/pkg/sync/time.go (hunk 26 above)
Comment (stevesloka) [reply discovery/pkg/sync/time.go]: yeah my bad, I got overzealous, I restored the old impl.

**Comment by stevesloka** [comment]
Testing tonight, I lost the golang metrics, looking to add those back in.

**Comment by stevesloka** [comment]
Ok latest commit adds those metrics back in.

File: discovery/pkg/openstack/translate.go (hunk 27 above)
Comment (alexbrand) [reply discovery/pkg/openstack/translate.go]: The alternative is to use the name + ID, but not sure if that is what we want.

Comment (rosskukulinski) [reply discovery/pkg/openstack/translate.go]: TBH I'm really not sure. I don't know how often folks name their load balancers the same thing in the same cluster.  Let's leave it as-is for now (at risk of duplicates).

If the LB doesn't have a name, do we put in the id for the label?

Comment (stevesloka) [reply discovery/pkg/openstack/translate.go]: yes name is tried first, then id is used.

File: discovery/cmd/openstack-discoverer/main.go [hunk 28]
```diff
@@ -71,6 +72,7 @@ func init() {
 	flag.IntVar(&prometheusListenPort, "prometheus-listen-address", 8080, "The address to listen on for Prometheus HTTP requests")
 	flag.Float64Var(&gimbalKubeClientQPS, "gimbal-client-qps", 5, "The maximum queries per second (QPS) that can be performed on the Gimbal Kubernetes API server")
 	flag.IntVar(&gimbalKubeClientBurst, "gimbal-client-burst", 10, "The maximum number of queries that can be performed on the Gimbal Kubernetes API server during a burst")
+	flag.DurationVar(&resyncInterval, "resync-interval", time.Minute*30, "Default resync period for watcher to refresh")
```
Comment (alexbrand) [inline discovery/cmd/openstack-discoverer/main.go]: Looks like this isn't used. I want to merge this PR to start testing, so I will open a new issue to discuss this flag. It seems like we can remove it, but wanted to confirm with Steve.

File: discovery/pkg/sync/queue.go [hunk 29]
```diff
@@ -110,17 +109,21 @@ func (sq *Queue) processNextWorkItem() bool {
...
 		sq.Logger.Errorf("got an unknown item of type %T in the queue", obj)
 		return true
 	}
 
-	err := action.Sync(sq.KubeClient, sq.Metrics, sq.BackendName)
+	err := action.Sync(sq.KubeClient, sq.Logger)
+	if err != nil {
+		action.SetMetricError(sq.Metrics)
+	}
+	action.SetMetrics(sq.KubeClient, sq.Metrics, sq.Logger)
```
Comment (alexbrand) [inline discovery/pkg/sync/queue.go]: I am wondering how these metrics will behave if we need to retry. Is it OK to set these metrics multiple times for the same service because of retries? Will open a new issue to discuss/look into this.
